### PR TITLE
deps/mobile: Bump `bazel` -> 6.3.1, `rules_java` -> 6.4.0, `io_bazel_rules_kotlin` -> 1.8

### PR DESCRIPTION
**Pull request by phlax**
Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
not sure if this will work and whether https://github.com/bazelbuild/rules_java/issues/64 is resolved - but essentially trying to unblock bazel upgrade

**Review by RyanTheOptimist — DISMISSED** [review]
Assuming this passes CI, LGTM. WOO HOO!!!

File: mobile/bazel/envoy_mobile_repositories.bzl [hunk 1]
```diff
@@ -65,10 +65,11 @@ def swift_repos():
...
         name = "rules_java",
-        sha256 = "19462d64b1586c0d4ea0e87f9325be2514f0eb84e56dbf3245450451b3701581",
-        strip_prefix = "rules_java-43243982abc76390ef64be62379a1353f9011771",
-        # TODO(jpsim): Switch back to bazelbuild repo when https://github.com/bazelbuild/rules_java/issues/64 is fixed
-        url = "https://github.com/jpsim/rules_java/archive/43243982abc76390ef64be62379a1353f9011771.tar.gz",
+        sha256 = "241822bf5fad614e3e1c42431002abd9af757136fa590a6a7870c6e0640a82e3",
+        strip_prefix = "rules_java-6.4.0",
+        url = "https://github.com/bazelbuild/rules_java/archive/6.4.0.tar.gz",
+        patch_args = ["-p1"],
+        patches = ["@envoy//bazel:rules_java.patch"],
```
Comment (keith) [inline mobile/bazel/envoy_mobile_repositories.bzl:72]: should we have a TODO for when we can drop this patch?

**Review by keith — APPROVED** [review]

**Comment by phlax** [comment]
@keith unfortunately its still failing mobile CI - we are now down to issues related to the rules_java upgade - any help resolving these would be much appreciated!

File: mobile/bazel/envoy_mobile_repositories.bzl (hunk 1 above)
Comment (phlax) [reply mobile/bazel/envoy_mobile_repositories.bzl:72]: i can add - when is that ?

Comment (keith) [reply mobile/bazel/envoy_mobile_repositories.bzl:72]: I'm not sure, i guess when that linked issue is fixed, or https://github.com/bazelbuild/bazel/issues/17085

**Review by keith — APPROVED** [review]

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**